File: todo_server.py
import asyncio
from typing import Optional, List
from pydantic import BaseModel
from mcp.server.fastmcp import FastMCP, Context

# Model Pydantic untuk struktur todo
class TodoItem(BaseModel):
    id: int
    title: str
    date: Optional[str] = None

# Simpan data todo di memory (list)
todos: List[TodoItem] = []

# Buat server FastMCP
mcp = FastMCP(name="TodoServer")

# Tool: tambah todo
@mcp.tool()
def add_todo(title: str, date: Optional[str] = None) -> TodoItem:
    new_id = len(todos) + 1
    todo = TodoItem(id=new_id, title=title, date=date)
    todos.append(todo)
    return todo

# Tool: list semua todo
@mcp.tool()
def list_todos() -> List[TodoItem]:
    return todos

# Tool: hapus todo berdasarkan id
@mcp.tool()
def delete_todo(todo_id: int) -> dict:
    global todos
    before = len(todos)
    todos = [t for t in todos if t.id != todo_id]
    after = len(todos)
    deleted = before - after
    return {"deleted": deleted}

# Tool: update judul todo
@mcp.tool()
def update_todo(todo_id: int, new_title: str) -> dict:
    for t in todos:
        if t.id == todo_id:
            t.title = new_title
            return {"message": "Todo updated", "todo": t}
    return {"error": "Todo not found"}

# Contoh tool: update bulk, ganti kata "aku" → "saya"
@mcp.tool()
def replace_aku_to_saya() -> dict:
    count = 0
    for t in todos:
        if "aku" in t.title:
            t.title = t.title.replace("aku", "saya")
            count += 1
    return {"replaced": count, "todos": todos}

# Jalankan server
if __name__ == "__main__":
    # Pakai transport stdio untuk development/testing
    mcp.run(transport="stdio")

File: todo_server.js
import { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import { StdioServerTransport } from "@modelcontextprotocol/sdk/server/stdio.js";
import { z } from "zod";

const TodoSchema = z.object({
  id: z.number().int().positive(),
  title: z.string(),
  date: z.string().optional().nullable(),
});

const mcpServer = new McpServer({
  name: "TodoServer",
  version: "0.1.0",
  description: "Todo list MCP server implemented in JavaScript",
});

const todos = [];
let nextId = 1;

mcpServer.registerTool("add_todo", {
  description: "Add a new todo item with title and optional date",
  // registerTool expects a raw Zod shape for inputSchema (not a z.object wrapper)
  inputSchema: {
    title: z.string().min(1, "Title is required"),
    date: z.string().min(1).optional(),
  }
}, async (args) => {
  const todo = {
    id: nextId++,
    title: args.title,
    date: args.date ?? null,
  };
  todos.push(todo);
  return todo;
});

mcpServer.registerTool("list_todos", {
  description: "List all todo items",
}, async () => todos);

mcpServer.registerTool("delete_todo", {
  description: "Delete a todo item by its ID",
  inputSchema: {
    todo_id: z.number().int().positive(),
  }
}, async (args) => {
  const before = todos.length;
  const index = todos.findIndex((todo) => todo.id === args.todo_id);
  if (index !== -1) {
    todos.splice(index, 1);
  }
  const after = todos.length;
  return { deleted: before - after };
});

mcpServer.registerTool("update_todo", {
  description: "Update the title of a todo item by its ID",
  inputSchema: {
    todo_id: z.number().int().positive(),
    new_title: z.string().min(1, "New title is required"),
  }
}, async (args) => {
  const todo = todos.find((item) => item.id === args.todo_id);
  if (!todo) {
    return { error: "Todo not found" };
  }
  todo.title = args.new_title;
  return { message: "Todo updated", todo };
});

mcpServer.registerTool("replace_aku_to_saya", {
  description: "Replace 'aku' with 'saya' in all todo titles",
}, async () => {
  let replaced = 0;
  for (const todo of todos) {
    if (todo.title.includes("aku")) {
      todo.title = todo.title.replace(/aku/g, "saya");
      replaced += 1;
    }
  }
  return { replaced, todos };
});

async function main() {
  const transport = new StdioServerTransport();
  await mcpServer.connect(transport);
}

main().catch((error) => {
  console.error("Failed to start TodoServer:", error);
  process.exit(1);
});
